pkg/virtctl/expose/expose.go

```diff
@@ -59,262 +61,232 @@ Possible types are (case insensitive, both single and plurant forms):
-	cmd.Flags().StringVar(&strIPFamilyPolicy, "ip-family-policy", "", "IP family policy defines whether the service can use IPv4, IPv6, or both. Valid values are 'SingleStack', 'PreferDualStack' or 'RequireDualStack'")
 	cmd.SetUsageTemplate(templates.UsageTemplate())
 
 	return cmd
 }
 
 func usage() string {
-	usage := `  # Expose SSH to a virtual machine instance called 'myvm' on each node via a NodePort service:
+	return `  # Expose SSH to a virtual machine instance called 'myvm' on each node via a NodePort service:
   {{ProgramName}} expose vmi myvm --port=22 --name=myvm-ssh --type=NodePort
 
   # Expose all defined pod-network ports of a virtual machine instance replicaset on a service:
   {{ProgramName}} expose vmirs myvmirs --name=vmirs-service
 
   # Expose port 8080 as port 80 from a virtual machine instance replicaset on a service:
   {{ProgramName}} expose vmirs myvmirs --port=80 --target-port=8080 --name=vmirs-service`
-	return usage
 }
 
-// executing the "expose" command
-func (o *Command) RunE(args []string) error {
+func (e *expose) run(cmd *cobra.Command, args []string) error {
 	// first argument is type of VM: VMI, VM or VMIRS
 	vmType := strings.ToLower(args[0])
 	// second argument must be name of the VM
 	vmName := args[1]
 
-	// these are used to convert the flag values into service spec values
-	var protocol v1.Protocol
-	var targetPort intstr.IntOrString
-	var serviceType v1.ServiceType
-
-	// convert from integer to the IntOrString type
-	targetPort = intstr.Parse(strTargetPort)
-
-	// convert from string to the protocol enum
-	switch strProtocol {
-	case "TCP":
-		protocol = v1.ProtocolTCP
-	case "UDP":
-		protocol = v1.ProtocolUDP
-	default:
-		return fmt.Errorf("unknown protocol: %s", strProtocol)
+	if err := e.parseFlags(); err != nil {
+		return err
 	}
 
-	// convert from string to the service type enum
-	switch strServiceType {
-	case "ClusterIP":
-		serviceType = v1.ServiceTypeClusterIP
-	case "NodePort":
-		serviceType = v1.ServiceTypeNodePort
-	case "LoadBalancer":
-		serviceType = v1.ServiceTypeLoadBalancer
-	case "ExternalName":
-		return fmt.Errorf("type: %s not supported", strServiceType)
-	default:
-		return fmt.Errorf("unknown service type: %s", strServiceType)
+	var err error
+	if e.namespace, _, err = e.clientConfig.Namespace(); err != nil {
+		return err
+	}
+	if e.client, err = kubecli.GetKubevirtClientFromClientConfig(e.clientConfig); err != nil {
+		return fmt.Errorf("cannot obtain KubeVirt client: %v", err)
 	}
 
-	ipFamilies, err := convertIPFamily(strIPFamily)
+	serviceSelector, ports, err := e.getServiceSelectorAndPorts(vmType, vmName)
 	if err != nil {
 		return err
 	}
+	if len(serviceSelector) == 0 {
+		return fmt.Errorf("cannot expose %s without any label: %s", vmType, vmName)
+	}
+	if len(ports) == 0 {
+		return fmt.Errorf("couldn't find port via --port flag or introspection")
+	}
```

Comment:
Why did you extract these checks out of `getServiceSelectorAndPorts`?

Reply:
Actually I did not (see original file L205), but you've got a good point. These can be part of `getServiceSelectorAndPorts`.